integration/cdi/src/main/java/org/apache/cxf/cdi/DelegateContextAnnotatedType.java

```diff
@@ -0,0 +1,182 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.cxf.cdi;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import static java.util.stream.Collectors.toSet;
+import javax.enterprise.inject.spi.AnnotatedConstructor;
+import javax.enterprise.inject.spi.AnnotatedField;
+import javax.enterprise.inject.spi.AnnotatedMethod;
+import javax.enterprise.inject.spi.AnnotatedType;
+import javax.enterprise.util.AnnotationLiteral;
+import javax.inject.Inject;
+import javax.ws.rs.core.Context;
+
+final class DelegateContextAnnotatedType<X> implements AnnotatedType<X> {
+    private static final Inject INJECT = new InjectLiteral();
+    private static final ContextResolved CONTEXT_RESOLVED = ContextResolved.LITERAL;
+    private final AnnotatedType<X> original;
+    private final Set<AnnotatedField<? super X>> replacedFields;
+
+    DelegateContextAnnotatedType(AnnotatedType<X> original) {
+        this.original = original;
+        this.replacedFields = replaceFields(original);
+    }
+
+    private Set<AnnotatedField<? super X>> replaceFields(AnnotatedType<? super X> delegate) {
+        return delegate.getFields().stream().map(this::wrap).collect(toSet());
+    }
+
+    Set<Type> getContextFieldTypes() {
+        return replacedFields.stream()
+                .filter(f -> f.isAnnotationPresent(Context.class) || f.isAnnotationPresent(ContextResolved.class))
```

Comment:
Wondering if we need to check for `ContextResolved` here? Technically, it should not be present, we add it along with inject, right?

Reply:
It would be for anything that isn't in the `STANDARD_CONTEXT` map, but is explicitly annotated and provided as a `Context` object.  Could be anything with a custom `ContextProvider` implemented that isn't injected using `@Context` but has the qualifier.

Reply:
Just to make sure the issue is understood, not sure if I'm not clear.  I'm registering a CDI bean per `@Context` object that can be injected.  I'm discovering them two ways:

- Any `@Context` injection point
- Any of the standard context objects, as defined by `InjectionUtils.STANDARD_CONTEXT_CLASSES`.

The latter is what enables `@Inject UriInfo` to work properly.  Without it, the injection point would not be satisfied.  The problem is that this works fine for the standard context classes, but not custom objects (e.g. `TracerContext`).  The closest I can do is if they have an injection point `@Inject @ContextResolved TracerContext` then I can still resolve that injection point, but I can't resolve `@Inject TracerContext` since I have no idea its supposed to be from the `@Context`.  

If we changed the behavior of these custom context objects so that they were added to a map in `InjectionUtils`, then I could resolve those classes as well and remove that extra check.

Reply:
Aha, thanks for explanation @johnament, the role of `@ContextResolved` becomes clear now. That's actually an interesting challenge to solve ... We definitely know all of them at the moment `JAXRSServerFactoryBean` is being constructed but it is too late with respect to CDI lifecycle (and what can we do at this point).

Reply:
So what do you think about adding support for registering the context type, manually?

Reply:
I think we could only manually register the context providers which come from CXF, but not the user ones. May be could deduct the context providers by introspecting the providers / resources / features during the `collect`? We have interface to check for, `ContextProvider`, this should for all cases. What do you think?

Reply:
We could, however it seems that the registration of the `ContextProvider` is done within CXF, which would be too late for CDI.  In addition, I don't want to require/expect that all `ContextProvider` impls are CDI beans.

Reply:
Sorry for keeping silence, haven't had time to get back to the issue. Spent some time playing with the context providers, my personal option (purely from user perspective), using `@Inject @ContextResolved SomeContext context` is not very intuitive.  

Getting back to the initial problem: `@Inject TracerContext` and `@Context TracerContext`. Besides the solution you have with `@ContextResolved`, we could also introduce 1) a marker interface for context objects 2) or an annotation (similar to `@ContextResolved`) for context objects 3) require context providers to be placed into CDI archives (than at least we could discover the possible context objects while processing annotated types) 

There are certainly more options available, but I think the 3rd one is as best as we could do without introducing too much of confusion. That would certainly impose some constraints (as you mentioned before) but may be it is not as bad as it sounds? What do you think?

Reply:
@reta  just want to clarify my point a bit more.

The simplest solution I can think of is to add this snippet to `BraveContextProvider`

```java
static {
        InjectionUtils.STANDARD_CONTEXT_CLASSES.add(BraveTracerContext.class.getName());
    }
```

Or do something like require a service loader to give the class name.  Then you wouldn't need to do things like make the providers CDI component classes (which doesn't sound right).

FWIW, I'm almost inclined to add `<beans bean-discovery-mode="none"/>` to modules that shouldn't have CDI beans within them, to avoid excessive classloadering.  But another day.

Reply:
Right, forgot about that, this is what you was referring to as manual registration. Ok, here is my suggestion, let us start from this simple solution (either using `InjectionUtils.STANDARD_CONTEXT_CLASSES`, or in safer with another collection, f.e. `InjectionUtils.NON_STANDARD_CONTEXT_CLASSES`), and revisit it later when the PR would be more or less feature complete. I suspect we are going to have quite a few challenges ahead, so we may use absolutely different approach at the end :-) At least we could get rid of `@ContextResolved` for now and move on. Sounds good?

Reply:
I would still recommend keeping `@ContextResolved` just to allow multiple beans to be registered (avoiding close tie in).  I'll work on getting the additional field processed and update this PR.

Reply:
Alright, so I've done some digging around and have bad news :-(

For the custom created context objects, pretty much everything fails CDI use cases.  In order to get the scoping correct, the context objects must be request scoped.  However, all of the custom context objects CXF has fail CDI requirements, they don't have default constructors.  As a result they can't be request scoped beans.

So each of those would need to remain as `@Context` objects for injection purposes.

Reply:
I expected the roadblocks here, `@Context` it be ...

Reply:
This is now working fine.  With the service loader approach, we can `@Inject` any context object we want now.